Improve bootstrap script

linuxbandit opened the pull request:
the ideal case is to use Make to run the target you need - no more hassle

The only limitation is when logs are called: if one wants to make logs only for a specific container, they have to use directly `helper.sh --logs containername1 containername2 ...`

Vagrant makes use of that makefile to bootstrap. Vagrant has now 3GB of memory since we have a bigger machine on the cloud.

Makefile also has function to help bumping the submodules at once: assuming all submodules are stable at any given commit, it should be safe to call it every now and then

linuxbandit committed:
Re-think script for the bootstrap

Also will be good for vagrant, which is in the works

linuxbandit committed:
Put better flags, correct minor stuff, add start.sh

linuxbandit committed:
Last finishments, bugfixes

serge1peshcoff commented:
can you update README to reflect these changes?

serge1peshcoff commented:
and probably specify here not only the commands that people need to run, but also some common commands, like executing a process in the container, getting log output etc.

linuxbandit commented:
Nice one about running the command in the container! Didn't think of it, I will add it

linuxbandit committed:
(Fix) Fix conditional for bump_repo

linuxbandit committed:
(Feat) Improve readme, add option for command execution

linuxbandit committed:
(Fix) .env is now copied from the template

linuxbandit committed:
(Impr) made 'make bump' idempotent(ish)

linuxbandit committed:
(Fix) Readme didn't display the table

linuxbandit commented:
Check it out now (easier to see [here](https://github.com/linuxbandit/oms-docker/blob/improve-bootstrap_script/README.md))

serge1peshcoff commented on README.md:
```diff
@@ -10,29 +10,29 @@ It makes use of docker, and docker-compose.
 
 ## Usage
-After running the system, you can navigate to it in your web browser. The URLs differ based on how you run it; however no matter how you decide to run it, _it is suggested_ to edit the `/etc/hosts` file to add the entry: `192.168.168.168 appserver traefik.appserver portainer.appserver`, to be able to use advanced features.
+After running the system, you can navigate to it in your web browser. The URLs differ based on how you run it; however no matter how you decide to run it, _it is suggested_ to edit the `/etc/hosts` file to add the entry: `192.168.168.168 my.appserver traefik.appserver portainer.appserver`, to be able to use advanced features.
```
1) I suggest putting other addresses as well (like kibana and my) to hosts
2) reflect here that on Windows it's `C:/Windows/system32/drivers/etc/hosts`

linuxbandit replied:
1. My is already there, but kibana not. I don't think others (e.g. status, and future ones) will be relevant
2. Yesterday I wanted to get it done for the simple case, though today I see that powershell has some more advanced features! I was thinking that cmd was shit (and it is), but ps is not as bad. I will add a .ps1 file that does the same

serge1peshcoff replied:
1. sorry, didn't notice. then it's all good

linuxbandit replied:
2. I wasted all my time until now, fuck windows. I will just include a batch file that must be right-clicked and run as admin, it opens the hosts file in notepad and the user has to paste and save, and invoke vagrant up manually. Fuck this shit

serge1peshcoff replied:
maybe just put 'to edit the `/etc/hosts` (on Windows, it's `C:/Windows/system32/drivers/etc/hosts`)' instead to Readme and let people figure that out by themselves?

serge1peshcoff commented on vagrant-post-script/orchestrate_docker.sh:
```diff
@@ -1,8 +1,8 @@
-./oms.sh up -d
\ No newline at end of file
+make bootstrap
+sleep 90 #to give time for the bootstrap
```
I suggest adding proper healthcheck here, something like

```
while (true) {
  if (fetch(oms_core_elixir_address)) {
    break;
  }

  sleep(2);
}`

(no idea how it would be in bash, so it is pseudo JS).
Not critical though, but would be nice, so we'd know if everything has really launched.

linuxbandit replied:
I wanted to avoid potentially infinite loops (if core is broken, e.g.)

Also, the address of frontend should be there, not of core? I guess curl is the real world implementation of what you suggested

serge1peshcoff replied:
1) we can add some threshold to exit it, e.g. 5 minutes
2) I'd say the backend. For example, the frontend is loaded but the backend is still loading (was the case for me at least, as somehow building oms-core-elixir takes a lot of time). Then the system will be still unusable till oms-core-elixir will start

linuxbandit replied:
I thought anyway that we should add some Healthcheck in the services (it was in my mind since long), to both know about their health and to accommodate situations like this one (and swarm, or consul, or portainer, or whatever like curl can know how to assess container's health).

Just a simple endpoint like `/ping` or `/health`

linuxbandit replied:
For now all it does is to check if I got the `<html>` tag, which is the error that Phoenix gives because it is not a known route

linuxbandit replied:
Also, I do that by changing the definition in `docker-compose.yml` of oms-core-elixir from `expose` to `ports`... i.e. I would need a PR on it now

serge1peshcoff commented on README.md:
```diff
@@ -42,39 +42,58 @@ For using it, it becomes:
-| 5432 | not registered | postgres |
 
 ## Deployment script
 The types of deployment:
```
can you add extra \n after this line? so Markdown would understand it's a list and display it properly

linuxbandit replied:
I don't understand, this part has been removed

serge1peshcoff replied:
no, the line after The types of deployment

linuxbandit committed:
(Impr) Now the script waits till oms-core is ready

Also add a sort-of helper for windows hosts
Remove useless script to make sure repo was with all submodules

linuxbandit committed:
(Fix) Fix readme

linuxbandit committed:
(Impr) Improve start.sh so it's idempotent

Can also be run without vagrant
Now before being launched, it checks if the hostfile is already good
